atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -371,7 +371,7 @@ public long setCommitIndex(final long commitIndex) {
   public void setLastApplied(final long lastApplied, final long lastAppliedTerm) {
     this.lastApplied = Math.max(this.lastApplied, lastApplied);
     this.lastAppliedTerm = Math.max(this.lastAppliedTerm, lastAppliedTerm);
-    if (state == State.ACTIVE) {
+    if (state == State.ACTIVE && this.lastApplied >= firstCommitIndex) {
```

Comment:
I can change it, if you want. My only caution is that this is a public method. And yes, currently it is only called from one place, but I wouldn't want to guarantee that for the future. Or maybe we can guarantee it, because it is such a narrow function.

Please advise whether I should change it or keep it as it is.

Reply:
That's a valid argument. Then to make it safe may be:
```
if (threadContext.isCurrentContext()) {
  // execute
} else {
// schedule
}
```

Reply:
I am ok with current way also.

Reply:
In that case, I would leave it as it is.

In the past I have been burned by code like 
```
if (threadContext.isCurrentContext()) {
  // execute
} else {
// schedule
}
```
Mostly, because it is hard to reason about code that sometimes gets executed synchronously and immediately, and sometimes gets executed asynchronously and at a deferred point in time.

I do like the idea, though. Maybe we can guarantee which thread we are on in the future.